### PR TITLE
Adds replica info to consensus snapshots

### LINKED ISSUE #933: [replication] Consensus: introduce shard state (dead or alive) and ops to change it
Peers need to know if shard replica is dead or alive.

Introduce:
1. Operations to change the state
2. Changes to consensus snapshot logic

### PR DESCRIPTION
**Pull request by e-ivkov**
Closes #933

### CODE REVIEW

File: lib/collection/src/collection_state.rs [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+pub enum ShardInfo {
+    ReplicaSet { replicas: HashMap<PeerId, IsActive> },
+    Single(PeerId),
+}
 
 #[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
 pub struct State {
     pub config: CollectionConfig,
-    pub shard_to_peer: HashMap<ShardId, PeerId>,
+    pub shards: HashMap<ShardId, ShardInfo>,
```
[the remark this answers is not in the document]
Comment (e-ivkov) [reply lib/collection/src/collection_state.rs:20]: It shouldn't, as we don't store the collection_state, we generate it on the flight to send as part of the consensus snapshot

File: lib/collection/src/shard/replica_set.rs [hunk 2]
```diff
@@ -18,23 +18,61 @@ pub type IsActive = bool;
...
+    ) -> CollectionResult<()> {
+        let removed_peers = self
+            .replica_state
+            .keys()
+            .filter(|peer_id| !replicas.contains_key(peer_id))
+            .copied()
+            .collect::<Vec<_>>();
+        for peer_id in removed_peers {
+            if peer_id == self.this_peer_id {
+                if let Some(shard) = &mut self.local {
```
Comment (agourlay) [inline lib/collection/src/shard/replica_set.rs:56]: Is it possible to get a `None` here?

File: lib/collection/src/shard/local_shard.rs [hunk 3]
```diff
@@ -563,4 +561,17 @@ impl Drop for LocalShard {
             }
         }
     }
+
+    pub async fn delete_from_disk(&self) -> CollectionResult<()> {
```
Comment (agourlay) [inline lib/collection/src/shard/local_shard.rs:565]: were you able to somehow test any of this part?

File: lib/collection/src/shard/replica_set.rs (hunk 2 above)
Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs:56]: It should not be possible generally, having it here would mean there was an error in logic somewhere

Comment (e-ivkov) [reply lib/collection/src/shard/replica_set.rs:56]: What do you think about adding a debug assert?

File: lib/collection/src/shard/local_shard.rs (hunk 3 above)
Comment (e-ivkov) [reply lib/collection/src/shard/local_shard.rs:565]: I can add a test if you think it makes sense?

File: lib/collection/src/shard/replica_set.rs (hunk 2 above)
Comment (agourlay) [reply lib/collection/src/shard/replica_set.rs:56]: I'd like a debug assert please :pray:

File: lib/collection/src/shard/local_shard.rs (hunk 3 above)
Comment (agourlay) [reply lib/collection/src/shard/local_shard.rs:565]: I'd love a test as a quick sanity check but only if the test is easy to write for you of course.

Comment (e-ivkov) [reply lib/collection/src/shard/local_shard.rs:565]: We decided to postpone the test to the next PR

**Review by agourlay — APPROVED** [review]
Looking forward to building on top of that :+1:

**Pull request merged**